cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,231 @@ struct ReplaceWithMaskFunctor {
     }
     return ReplaceWithMask<Type>::ExecArrayMask(ctx, array, mask, replacements, output);
   }
+
+  static std::shared_ptr<KernelSignature> GetParameters(detail::GetTypeId get_id){
+    return KernelSignature::Make(
+        {InputType::Array(get_id.id), InputType(boolean()), InputType(get_id.id)},
+        OutputType(FirstType));
+  }
 };
 
-}  // namespace
+template <typename Type>
+void FillNullInDirectionImpl(const ArrayData& array, const uint8_t* null_bitmap,
+                             ArrayData* output, int8_t direction) {
+  uint8_t* out_bitmap = output->buffers[0]->mutable_data();
+  uint8_t* out_values = output->buffers[1]->mutable_data();
+  arrow::internal::CopyBitmap(array.buffers[0]->data(), array.offset, array.length,
+                              out_bitmap, output->offset);
+  ReplaceWithMask<Type>::CopyData(*array.type, out_values,
+                                  /*out_offset=*/0, array, /*in_offset=*/0, array.length);
 
-const FunctionDoc replace_with_mask_doc(
-    "Replace items selected with a mask",
-    ("Given an array and a boolean mask (either scalar or of equal length),\n"
-     "along with replacement values (either scalar or array),\n"
-     "each element of the array for which the corresponding mask element is\n"
-     "true will be replaced by the next value from the replacements,\n"
-     "or with null if the mask is null.\n"
-     "Hence, for replacement arrays, len(replacements) == sum(mask == true)."),
-    {"values", "mask", "replacements"});
+  auto array_scalars = arrow::MakeArray(array.Copy());
 
-void RegisterVectorReplace(FunctionRegistry* registry) {
-  auto func = std::make_shared<VectorFunction>("replace_with_mask", Arity::Ternary(),
-                                               &replace_with_mask_doc);
+  int64_t write_offset = direction == 1 ? 0 : array.length - 1;
+  int64_t bitmap_offset = 0;
+  int64_t current_value_offset = 0;
+  bool has_fill_value = false;
+  arrow::internal::OptionalBinaryBitBlockCounter counter(
+      null_bitmap, array.offset, null_bitmap, array.offset, array.length);
+
+  while (write_offset < array.length && write_offset >= 0) {
+    BitBlockCount block = counter.NextAndBlock();
+    if (block.AllSet()) {
+      current_value_offset = write_offset + direction * (block.length - 1);
```

Comment:
has_fill_value = true?

Reply:
We should also have a unit test that covers this case.

Reply:
And this should update bitmap_offset. (Or really, it should be updated at the bottom along with write_offset.)